tools/c7n_azure/c7n_azure/resources/storage.py

```diff
@@ -92,65 +114,106 @@ class StorageSetNetworkRulesAction(AzureBaseAction):
-                for r in self.data['ip-rules']]
-
-        if 'virtual-network-rules' in self.data:
-            rule_set.virtual_network_rules = [
-                VirtualNetworkRule(
-                    virtual_network_resource_id=r['virtual-network-resource-id'],
-                    action='Allow')  # 'Allow' is the only allowed action
-                for r in self.data['virtual-network-rules']]
-
-        if len(self.data.get('bypass', [])) > 0:
-            rule_set.bypass = ','.join(self.data['bypass'])
-        else:
-            rule_set.bypass = 'None'
+        rules = self._build_ip_rules(resource, self.data.get('ip-rules', []))
+
+        # Build out the ruleset model to update the resource
+        rule_set = NetworkRuleSet(default_action=self.data.get('default-action', 'Deny'))
+
+        # If the user has too many rules log and skip
+        if len(rules) > self.rule_limit:
+            self._log.error("Skipped updating firewall for %s. "
+                            "%s exceeds maximum rule count of %s." %
+                            (resource['name'], len(rules), self.rule_limit))
+            return
 
+        # Add IP rules
+        rule_set.ip_rules = [IPRule(ip_address_or_range=r) for r in rules]
+
+        # Add VNET rules
+        vnet_rules = self._build_vnet_rules(resource, self.data.get('virtual-network-rules', []))
+        rule_set.virtual_network_rules = [
+            VirtualNetworkRule(virtual_network_resource_id=r) for r in vnet_rules]
+
+        # Configure BYPASS
+        bypass_rules = self._build_bypass_rules(resource, self.data.get('bypass', []))
+        rule_set.bypass = ','.join(bypass_rules or ['None'])
+
+        # Update resource
         self.client.storage_accounts.update(
             resource['resourceGroup'],
             resource['name'],
             StorageAccountUpdateParameters(network_rule_set=rule_set))
 
+    def _build_bypass_rules(self, resource, new_rules):
+        if self.append:
+            existing_bypass = resource['properties']['networkAcls'].get('bypass', '').split(',')
+            without_duplicates = [r for r in existing_bypass if r not in new_rules]
+            new_rules.extend(without_duplicates)
+        return new_rules
+
+    def _build_vnet_rules(self, resource, new_rules):
+        if self.append:
+            existing_rules = [r['id'] for r in
+                              resource['properties']['networkAcls'].get('virtualNetworkRules', [])]
+            without_duplicates = [r for r in existing_rules if r not in new_rules]
+            new_rules.extend(without_duplicates)
+        return new_rules
+
+    def _build_ip_rules(self, resource, new_rules):
+        rules = []
+        for rule in new_rules:
+            resolved_set = resolve_service_tag_alias(rule)
+            if resolved_set:
+                ranges = list(resolved_set.iter_cidrs())
+                for r in range(len(ranges)):
+                    if len(ranges[r]) == 1:
```

Comment:
What if it is 2?

Reply:
Wait, are you aware of this obscure storage firewall behavior where they block /31 blocks? Because that would be weird unless I told you about it :)   

They do actually block /31 and /32 (inexplicably, and only on storage firewalls), so technically we should have some advanced handling for /31 to convert it to a range syntax, but its sort of tricky to do and service tags don't contain any /31's so it isn't really needed.  

But I find it surprising that you knew that I skipped this, lol